File: ThreeDSecure/src/test/java/com/braintreepayments/api/CardinalClientUnitTest.java
package com.braintreepayments.api;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertSame;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Matchers.eq;
import static org.mockito.Matchers.isNull;
import static org.mockito.Matchers.same;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static org.powermock.api.mockito.PowerMockito.mockStatic;

import android.app.Activity;
import android.content.Context;

import androidx.fragment.app.FragmentActivity;

import com.cardinalcommerce.cardinalmobilesdk.Cardinal;
import com.cardinalcommerce.cardinalmobilesdk.enums.CardinalEnvironment;
import com.cardinalcommerce.cardinalmobilesdk.models.CardinalConfigurationParameters;
import com.cardinalcommerce.cardinalmobilesdk.services.CardinalInitService;
import com.cardinalcommerce.cardinalmobilesdk.services.CardinalValidateReceiver;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

@RunWith(PowerMockRunner.class)
@PrepareForTest({Cardinal.class})
public class CardinalClientUnitTest {

    private Cardinal cardinalInstance;

    private FragmentActivity activity;
    private Configuration configuration;

    private Context context;
    private CardinalInitializeCallback cardinalInitializeCallback;
    private CardinalValidateReceiver cardinalValidateReceiver;

    @Before
    public void beforeEach() {
        mockStatic(Cardinal.class);
        context = mock(Context.class);
        configuration = mock(Configuration.class);
        cardinalInitializeCallback = mock(CardinalInitializeCallback.class);
        cardinalInstance = mock(Cardinal.class);

        activity = mock(FragmentActivity.class);
        cardinalValidateReceiver = mock(CardinalValidateReceiver.class);
    }

    @Test
    public void initialize_configuresDefaultCardinalConfigurationParameters() throws BraintreeException {
        when(Cardinal.getInstance()).thenReturn(cardinalInstance);
        CardinalClient sut = new CardinalClient();

        ThreeDSecureRequest request = new ThreeDSecureRequest();
        sut.initialize(context, configuration, request, cardinalInitializeCallback);

        ArgumentCaptor<CardinalConfigurationParameters> captor = ArgumentCaptor.forClass(CardinalConfigurationParameters.class);
        verify(cardinalInstance).configure(same(context), captor.capture());

        CardinalConfigurationParameters parameters = captor.getValue();
        assertEquals(CardinalEnvironment.STAGING, parameters.getEnvironment());
        assertEquals(8000, parameters.getRequestTimeout());
        assertTrue(parameters.isEnableDFSync());
    }

    @Test
    public void initialize_whenV2UiCustomizationNotNull_setsCardinalConfigurationParameters() throws BraintreeException {
        when(Cardinal.getInstance()).thenReturn(cardinalInstance);
        CardinalClient sut = new CardinalClient();

        ThreeDSecureV2UiCustomization v2UiCustomization = new ThreeDSecureV2UiCustomization();

        ThreeDSecureRequest request = new ThreeDSecureRequest();
        request.setV2UiCustomization(v2UiCustomization);
        sut.initialize(context, configuration, request, cardinalInitializeCallback);

        ArgumentCaptor<CardinalConfigurationParameters> captor = ArgumentCaptor.forClass(CardinalConfigurationParameters.class);
        verify(cardinalInstance).configure(same(context), captor.capture());

        CardinalConfigurationParameters parameters = captor.getValue();
        assertEquals(request.getV2UiCustomization().getCardinalUiCustomization(), parameters.getUICustomization());
    }

    @Test
    public void initialize_whenEnvironmentProduction_configuresCardinalEnvironmentProduction() throws BraintreeException {
        when(Cardinal.getInstance()).thenReturn(cardinalInstance);
        when(configuration.getEnvironment()).thenReturn("production");

        CardinalClient sut = new CardinalClient();

        ThreeDSecureRequest request = new ThreeDSecureRequest();
        sut.initialize(context, configuration, request, cardinalInitializeCallback);

        ArgumentCaptor<CardinalConfigurationParameters> captor = ArgumentCaptor.forClass(CardinalConfigurationParameters.class);
        verify(cardinalInstance).configure(same(context), captor.capture());

        CardinalConfigurationParameters parameters = captor.getValue();
        assertEquals(CardinalEnvironment.PRODUCTION, parameters.getEnvironment());
    }

    @Test
    public void initialize_returnsConsumerSessionIdToListener() throws BraintreeException {
        when(Cardinal.getInstance()).thenReturn(cardinalInstance);
        when(configuration.getCardinalAuthenticationJwt()).thenReturn("token");
        CardinalClient sut = new CardinalClient();

        ThreeDSecureRequest request = new ThreeDSecureRequest();
        sut.initialize(context, configuration, request, cardinalInitializeCallback);

        ArgumentCaptor<CardinalInitService> captor = ArgumentCaptor.forClass(CardinalInitService.class);
        verify(cardinalInstance).init(eq("token"), captor.capture());

        CardinalInitService cardinalInitService = captor.getValue();
        cardinalInitService.onSetupCompleted("session-id");

        verify(cardinalInitializeCallback).onResult(sut.getConsumerSessionId(), null);
        assertEquals("session-id", sut.getConsumerSessionId());
    }

    @Test
    public void initialize_whenConsumerSessionIdIsNull_returnsBraintreeExceptionToListener() throws BraintreeException {
        when(Cardinal.getInstance()).thenReturn(cardinalInstance);
        when(configuration.getCardinalAuthenticationJwt()).thenReturn("token");
        CardinalClient sut = new CardinalClient();

        ThreeDSecureRequest request = new ThreeDSecureRequest();
        sut.initialize(context, configuration, request, cardinalInitializeCallback);

        ArgumentCaptor<CardinalInitService> captor = ArgumentCaptor.forClass(CardinalInitService.class);
        verify(cardinalInstance).init(eq("token"), captor.capture());

        CardinalInitService cardinalInitService = captor.getValue();
        cardinalInitService.onValidated(null, null);

        ArgumentCaptor<BraintreeException> exceptionCaptor = ArgumentCaptor.forClass(BraintreeException.class);
        verify(cardinalInitializeCallback).onResult(isNull(String.class), exceptionCaptor.capture());
        assertEquals(exceptionCaptor.getValue().getMessage(), "consumer session id not available");
    }

    @Test
    public void initialize_onCardinalConfigureRuntimeException_throwsError() {
        when(Cardinal.getInstance()).thenReturn(cardinalInstance);
        when(configuration.getCardinalAuthenticationJwt()).thenReturn("token");
        CardinalClient sut = new CardinalClient();

        RuntimeException runtimeException = new RuntimeException("fake message");
        doThrow(runtimeException)
                .when(cardinalInstance)
                .configure(any(Context.class), any(CardinalConfigurationParameters.class));

        ThreeDSecureRequest request = new ThreeDSecureRequest();

        try {
            sut.initialize(context, configuration, request, cardinalInitializeCallback);
            fail("should not get here");
        } catch (BraintreeException e) {
            assertEquals("Cardinal SDK configure Error.", e.getMessage());
            assertSame(runtimeException, e.getCause());
        }
    }

    @Test
    public void initialize_onCardinalInitRuntimeException_throwsError() {
        when(Cardinal.getInstance()).thenReturn(cardinalInstance);
        when(configuration.getCardinalAuthenticationJwt()).thenReturn("token");
        CardinalClient sut = new CardinalClient();

        RuntimeException runtimeException = new RuntimeException("fake message");
        doThrow(runtimeException)
                .when(cardinalInstance)
                .init(anyString(), any(CardinalInitService.class));

        ThreeDSecureRequest request = new ThreeDSecureRequest();
        try {
            sut.initialize(context, configuration, request, cardinalInitializeCallback);
            fail("should not get here");
        } catch (BraintreeException e) {
            assertEquals("Cardinal SDK init Error.", e.getMessage());
            assertSame(runtimeException, e.getCause());
        }
    }

    @Test
    public void continueLookup_continuesCardinalLookup() throws BraintreeException {
        when(Cardinal.getInstance()).thenReturn(cardinalInstance);
        CardinalClient sut = new CardinalClient();

        ThreeDSecureLookup threeDSecureLookup = mock(ThreeDSecureLookup.class);
        when(threeDSecureLookup.getTransactionId()).thenReturn("sample-transaction-id");
        when(threeDSecureLookup.getPareq()).thenReturn("sample-payer-authentication-request");

        ThreeDSecureResult threeDSecureResult = mock(ThreeDSecureResult.class);
        when(threeDSecureResult.getLookup()).thenReturn(threeDSecureLookup);

        sut.continueLookup(activity, threeDSecureResult, cardinalValidateReceiver);

        verify(cardinalInstance).cca_continue(
                "sample-transaction-id",
                "sample-payer-authentication-request",
                activity,
                cardinalValidateReceiver
        );
    }

    @Test
    public void continueLookup_onCardinalRuntimeException_throwsError() {
        when(Cardinal.getInstance()).thenReturn(cardinalInstance);
        RuntimeException runtimeException = new RuntimeException("fake message");
        doThrow(runtimeException)
                .when(cardinalInstance).cca_continue(anyString(), anyString(), any(Activity.class), any(CardinalValidateReceiver.class));
        CardinalClient sut = new CardinalClient();

        ThreeDSecureLookup threeDSecureLookup = mock(ThreeDSecureLookup.class);
        when(threeDSecureLookup.getTransactionId()).thenReturn("sample-transaction-id");
        when(threeDSecureLookup.getPareq()).thenReturn("sample-payer-authentication-request");

        ThreeDSecureResult threeDSecureResult = mock(ThreeDSecureResult.class);
        when(threeDSecureResult.getLookup()).thenReturn(threeDSecureLookup);

        try {
            sut.continueLookup(activity, threeDSecureResult, cardinalValidateReceiver);
            fail("should not get here");
        } catch (BraintreeException e) {
            assertEquals("Cardinal SDK cca_continue Error.", e.getMessage());
            assertSame(runtimeException, e.getCause());
        }
    }
}

File: ThreeDSecure/src/test/java/com/braintreepayments/api/threedsecure/CardinalClientUnitTest.kt
package com.braintreepayments.api.threedsecure

import android.content.Context
import androidx.fragment.app.FragmentActivity
import com.braintreepayments.api.core.BraintreeException
import com.braintreepayments.api.core.Configuration
import com.cardinalcommerce.cardinalmobilesdk.Cardinal
import com.cardinalcommerce.cardinalmobilesdk.enums.CardinalEnvironment
import com.cardinalcommerce.cardinalmobilesdk.enums.CardinalUiType
import com.cardinalcommerce.cardinalmobilesdk.models.CardinalChallengeObserver
import com.cardinalcommerce.cardinalmobilesdk.models.CardinalConfigurationParameters
import com.cardinalcommerce.cardinalmobilesdk.services.CardinalInitService
import com.cardinalcommerce.cardinalmobilesdk.services.CardinalValidateReceiver
import io.mockk.every
import io.mockk.mockk
import io.mockk.mockkStatic
import io.mockk.slot
import io.mockk.verify
import junit.framework.TestCase.assertEquals
import junit.framework.TestCase.assertSame
import junit.framework.TestCase.assertTrue
import junit.framework.TestCase.fail
import org.junit.Before
import org.junit.Test

class CardinalClientUnitTest {

    private lateinit var cardinalInstance: Cardinal
    private lateinit var activity: FragmentActivity
    private lateinit var configuration: Configuration
    private lateinit var context: Context
    private lateinit var cardinalInitializeCallback: CardinalInitializeCallback
    private lateinit var cardinalValidateReceiver: CardinalValidateReceiver
    private lateinit var cardinalChallengeObserver: CardinalChallengeObserver

    @Before
    fun beforeEach() {
        mockkStatic(Cardinal::class)

        context = mockk(relaxed = true)
        configuration = mockk(relaxed = true)
        cardinalInitializeCallback = mockk(relaxed = true)
        cardinalInstance = mockk(relaxed = true)
        activity = mockk(relaxed = true)
        cardinalValidateReceiver = mockk(relaxed = true)
        cardinalChallengeObserver = mockk(relaxed = true)
    }

    @Test
    @Throws(BraintreeException::class)
    fun initialize_configuresDefaultCardinalConfigurationParameters() {
        every { Cardinal.getInstance() } returns cardinalInstance

        val sut = CardinalClient()
        val request = ThreeDSecureRequest()
        sut.initialize(context, configuration, request, cardinalInitializeCallback)

        val parametersSlot = slot<CardinalConfigurationParameters>()
        verify { cardinalInstance.configure(context, capture(parametersSlot)) }

        val parameters = parametersSlot.captured
        assertEquals(CardinalEnvironment.STAGING, parameters.environment)
        assertEquals(8000, parameters.requestTimeout.toLong())
        assertTrue(parameters.isEnableDFSync)
    }

    @Test
    @Throws(BraintreeException::class)
    fun initialize_whenV2UiCustomizationNotNull_setsCardinalConfigurationParameters() {
        every { Cardinal.getInstance() } returns cardinalInstance

        val sut = CardinalClient()
        val v2UiCustomization = ThreeDSecureV2UiCustomization()
        val request = ThreeDSecureRequest()
        request.v2UiCustomization = v2UiCustomization
        sut.initialize(context, configuration, request, cardinalInitializeCallback)

        val parametersSlot = slot<CardinalConfigurationParameters>()
        verify { cardinalInstance.configure(context, capture(parametersSlot)) }

        val parameters = parametersSlot.captured
        assertEquals(
            request.v2UiCustomization!!.cardinalUiCustomization.toolbarCustomization,
            parameters.uiCustomization.toolbarCustomization
        )
    }

    @Test
    @Throws(BraintreeException::class)
    fun initialize_whenEnvironmentProduction_configuresCardinalEnvironmentProduction() {
        every { Cardinal.getInstance() } returns cardinalInstance
        every { configuration.environment } returns "production"

        val sut = CardinalClient()
        val request = ThreeDSecureRequest()
        sut.initialize(context, configuration, request, cardinalInitializeCallback)

        val parametersSlot = slot<CardinalConfigurationParameters>()
        verify { cardinalInstance.configure(context, capture(parametersSlot)) }

        val parameters = parametersSlot.captured
        assertEquals(CardinalEnvironment.PRODUCTION, parameters.environment)
    }

    @Test
    @Throws(BraintreeException::class)
    fun initialize_whenUiTypeNotNull_setsCardinalConfigurationParameters() {
        every { Cardinal.getInstance() } returns cardinalInstance

        val sut = CardinalClient()
        val request = ThreeDSecureRequest().apply {
            uiType = ThreeDSecureUiType.BOTH
        }
        sut.initialize(context, configuration, request, cardinalInitializeCallback)

        val parametersSlot = slot<CardinalConfigurationParameters>()
        verify { cardinalInstance.configure(context, capture(parametersSlot)) }

        val parameters = parametersSlot.captured
        assertEquals(CardinalUiType.BOTH, parameters.uiType)
    }

    @Test
    @Throws(BraintreeException::class)
    fun initialize_whenRenderTypeNotNull_setsCardinalConfigurationParameters() {
        every { Cardinal.getInstance() } returns cardinalInstance

        val sut = CardinalClient()
        val request = ThreeDSecureRequest().apply {
            renderTypes = listOf(
                ThreeDSecureRenderType.OTP,
                ThreeDSecureRenderType.SINGLE_SELECT,
                ThreeDSecureRenderType.MULTI_SELECT,
                ThreeDSecureRenderType.OOB,
                ThreeDSecureRenderType.RENDER_HTML,
            )
        }
        sut.initialize(context, configuration, request, cardinalInitializeCallback)

        val parametersSlot = slot<CardinalConfigurationParameters>()
        verify { cardinalInstance.configure(context, capture(parametersSlot)) }

        val parameters = parametersSlot.captured
        assertEquals(5, parameters.renderType.length())
    }

    @Test
    @Throws(BraintreeException::class)
    fun initialize_returnsConsumerSessionIdToListener() {
        every { Cardinal.getInstance() } returns cardinalInstance
        every { configuration.cardinalAuthenticationJwt } returns "token"

        val sut = CardinalClient()
        val request = ThreeDSecureRequest()
        sut.initialize(context, configuration, request, cardinalInitializeCallback)

        val cardinalInitServiceSlot = slot<CardinalInitService>()
        verify { cardinalInstance.init("token", capture(cardinalInitServiceSlot)) }

        val cardinalInitService = cardinalInitServiceSlot.captured
        cardinalInitService.onSetupCompleted("session-id")

        verify { cardinalInitializeCallback.onResult(sut.consumerSessionId, null) }
        assertEquals("session-id", sut.consumerSessionId)
    }

    @Test
    @Throws(BraintreeException::class)
    fun initialize_whenConsumerSessionIdIsNull_returnsBraintreeExceptionToListener() {
        every { Cardinal.getInstance() } returns cardinalInstance
        every { configuration.cardinalAuthenticationJwt } returns "token"

        val sut = CardinalClient()
        val request = ThreeDSecureRequest()
        sut.initialize(context, configuration, request, cardinalInitializeCallback)

        val cardinalInitServiceSlot = slot<CardinalInitService>()
        verify { cardinalInstance.init("token", capture(cardinalInitServiceSlot)) }

        val cardinalInitService = cardinalInitServiceSlot.captured
        cardinalInitService.onValidated(null, null)

        val exceptionSlot = slot<BraintreeException>()
        verify { cardinalInitializeCallback.onResult(isNull(), capture(exceptionSlot)) }
        assertEquals(exceptionSlot.captured.message, "consumer session id not available")
    }


    @Test
    fun `when cardinal configuration is called with a requestorAppUrl, sets threeDSRequestorAppURL`() {
        every { Cardinal.getInstance() } returns cardinalInstance

        val sut = CardinalClient()
        val request = ThreeDSecureRequest()
        request.requestorAppUrl = "www.paypal.com"
        sut.initialize(context, configuration, request, cardinalInitializeCallback)

        val parametersSlot = slot<CardinalConfigurationParameters>()
        verify { cardinalInstance.configure(context, capture(parametersSlot)) }

        val parameters = parametersSlot.captured
        assertEquals("www.paypal.com", parameters.threeDSRequestorAppURL)
    }

    @Test
    fun initialize_onCardinalConfigureRuntimeException_throwsError() {
        every { Cardinal.getInstance() } returns cardinalInstance
        every { configuration.cardinalAuthenticationJwt } returns "token"

        val sut = CardinalClient()
        val runtimeException = RuntimeException("fake message")
        every { cardinalInstance.configure(any(), any()) } throws runtimeException

        val request = ThreeDSecureRequest()
        try {
            sut.initialize(context, configuration, request, cardinalInitializeCallback)
            fail("should not get here")
        } catch (e: BraintreeException) {
            assertEquals("Cardinal SDK configure Error.", e.message)
            assertSame(runtimeException, e.cause)
        }
    }

    @Test
    fun initialize_onCardinalInitRuntimeException_throwsError() {
        every { Cardinal.getInstance() } returns cardinalInstance
        every { configuration.cardinalAuthenticationJwt } returns "token"

        val sut = CardinalClient()
        val runtimeException = RuntimeException("fake message")
        every { cardinalInstance.init(any(), any()) } throws runtimeException

        val request = ThreeDSecureRequest()
        try {
            sut.initialize(context, configuration, request, cardinalInitializeCallback)
            fail("should not get here")
        } catch (e: BraintreeException) {
            assertEquals("Cardinal SDK init Error.", e.message)
            assertSame(runtimeException, e.cause)
        }
    }

    @Test
    @Throws(BraintreeException::class)
    fun continueLookup_continuesCardinalLookup() {
        every { Cardinal.getInstance() } returns cardinalInstance

        val sut = CardinalClient()
        val threeDSecureLookup = mockk<ThreeDSecureLookup>(relaxed = true)

        every { threeDSecureLookup.transactionId } returns "sample-transaction-id"
        every { threeDSecureLookup.pareq } returns "sample-payer-authentication-request"

        val threeDSecureParams = mockk<ThreeDSecureParams>(relaxed = true)
        every { threeDSecureParams.lookup } returns threeDSecureLookup

        sut.continueLookup(threeDSecureParams, cardinalChallengeObserver)
        verify {
            cardinalInstance.cca_continue(
                "sample-transaction-id",
                "sample-payer-authentication-request",
                cardinalChallengeObserver
            )
        }
    }

    @Test
    fun continueLookup_onCardinalRuntimeException_throwsError() {
        every { Cardinal.getInstance() } returns cardinalInstance

        val runtimeException = RuntimeException("fake message")
        every { cardinalInstance.cca_continue(any(), any(), any()) } throws runtimeException

        val sut = CardinalClient()
        val threeDSecureLookup = mockk<ThreeDSecureLookup>(relaxed = true)
        every { threeDSecureLookup.transactionId } returns "sample-transaction-id"
        every { threeDSecureLookup.pareq } returns "sample-payer-authentication-request"

        val threeDSecureParams = mockk<ThreeDSecureParams>(relaxed = true)
        every { threeDSecureParams.lookup } returns threeDSecureLookup

        try {
            sut.continueLookup(threeDSecureParams, cardinalChallengeObserver)
            fail("should not get here")
        } catch (e: BraintreeException) {
            assertEquals("Cardinal SDK cca_continue Error.", e.message)
            assertSame(runtimeException, e.cause)
        }
    }

    @Test
    fun `when continueLookup is called with a null challengeObserver, throws BraintreeException`() {
        val sut = CardinalClient()

        try {
            sut.continueLookup(mockk(), null)
        } catch (e: BraintreeException) {
            assertEquals("challengeObserver is null", e.message)
        }
    }

    @Test
    fun cleanup_cleansUpCardinalInstance() {
        every { Cardinal.getInstance() } returns cardinalInstance

        val sut = CardinalClient()
        sut.cleanup()
        verify { cardinalInstance.cleanup() }
    }
}
